Add a shaping example

wez opened the pull request:
I'm submitting this partially because I think it is a more complete
shaping example, partially because I'd value your feedback on
whether it is a correct way to do things, and partially to draw you
attention to a couple of things that could be streamlined!

--

Use `cargo run --example shape` to execute it.  It assumes that
is has access to some fonts that are present on my system.

This example demonstrates a couple of things:

* How to avoid so many of the unwrap calls that are present in the
  other examples(!)
* How to use transmute to manage the lifetimes so that the
  a set of related objects can be held together in a struct
* How to deal with `Info::init_from_glyphs` treating unknown glyphs
  as ZWNJ chars so that we can process font fallback to eg: emoji.

Refs: https://github.com/wez/wezterm/issues/66
Refs: https://github.com/yeslogic/allsorts/issues/8

wez committed:
Add a shaping example

Use `cargo run --example shape` to execute it.  It assumes that
is has access to some fonts that are present on my system.

This example demonstrates a couple of things:

* How to avoid so many of the unwrap calls that are present in the
  other examples(!)
* How to use transmute to work manage the lifetimes so that the
  a set of related objects can be held together in a struct
* How to deal with `Info::init_from_glyphs` treating unknown glyphs
  as ZWNJ chars so that we can process font fallback to eg: emoji.

Refs: https://github.com/wez/wezterm/issues/66
Refs: https://github.com/yeslogic/allsorts/issues/8

wez commented on examples/shape.rs:
```diff
@@ -0,0 +1,344 @@
+        // That in turn allows us to store all of these derived items
+        // into a struct and manage their lifetimes together.
+        let file: OpenTypeFile<'static> =
+            unsafe { transmute(owned_scope.scope().read::<OpenTypeFile>()?) };
```
I don't like this transmute: it would be great if there were a clear and easy way to end up with an owned instance so that the lifetimes are easier to deal with for the consuming code.

wez commented on examples/shape.rs:
```diff
@@ -0,0 +1,344 @@
+            &mut glyphs,
+        )?;
+
+        // Note: init_from_glyphs silently elides entries that
```
This behavior was surprising; it would be great if there was a reasonable way to signal that the glyph was unknown rather than stripping it out here:

https://docs.rs/allsorts/0.1.0/src/allsorts/gpos.rs.html#292

CryZe commented on Cargo.toml:
```diff
@@ -16,3 +16,4 @@ encoding_rs = "0.8.16"
 getopts = "0.2"
 gumdrop = "0.7.0"
 itertools = "0.8"
+anyhow = "1.0"
```
Shouldn't this be a dev-dependency?

wez committed:
various improvements

add some comments, more useful advance metrics, improve running out of
fallback fonts, move anyhow to dev-dependencies
